tools/web/index.ts

```diff
@@ -1,185 +1,399 @@
+                        ResultTree.nestingMethod = NestingMethod.ByPath;
+                        reloadResultTree();
+                        b.invalidateStyles();
+                        b.invalidate();
+                    }
+                },
+                [
+                    styles.switchButton,
+                    ResultTree.nestingMethod === NestingMethod.ByDescribe ? styles.disabledButton : styles.activeButton
+                ]
+            )
+        ],
+        styles.buttonGroup
+    );
 }
 
-function createActionUI(action: s.IAction): b.IBobrilChildren {
-    switch (action.type) {
-        case "command": {
-            const command = <s.IActionCommand>action;
-            return createButton(command.name, command.enabled, () => {
-                com.runAction(command.id);
-            })
-        }
-        case "combo": {
-            const combo = <s.IActionCombo>action;
-            return [combo.label, " ", createCombo(combo.selected, combo.options, (id) => {
-                com.runAction(id);
-            })];
-        }
-        default: {
-            return "Unknown action type " + action.type;
-        }
+function createResultTree(): b.IBobrilChildren {
+    if (selectedAgentIndex !== lastLoadedAgentIndex || lastLoadedDataVersion !== s.testSvrDataVersion) {
+        reloadResultTree();
     }
+
+    return b.styledDiv(resultTree.toComponent(), styles.results);
 }
 
+function reloadResultTree() {
+    let separatedResults: testReportAnalyzer.SeparatedTests = testReportAnalyzer.analyze(
+        s.testSvrState.agents[selectedAgentIndex].nested
+    );
+
+    resultTree.reloadSOTs(separatedResults);
+
+    lastLoadedAgentIndex = selectedAgentIndex;
+    lastLoadedDataVersion = s.testSvrDataVersion;
+
+    b.invalidate();
+}
+
+let selectedAgentIndex = -1;
+let lastLoadedAgentIndex = -1;
+
+let lastLoadedDataVersion = 0;
+
+const resultTree: ResultTree = new ResultTree();
+
 com.reconnect();
 
-let selectedAgent = -1;
 b.init(() => {
-    if (selectedAgent >= s.testSvrState.agents.length) {
-        selectedAgent = -1;
+    if (selectedAgentIndex >= s.testSvrState.agents.length) {
+        selectedAgentIndex = -1;
     }
-    if (selectedAgent === -1 && s.testSvrState.agents.length > 0) {
-        selectedAgent = 0;
+    if (selectedAgentIndex === -1 && s.testSvrState.agents.length > 0) {
+        selectedAgentIndex = 0;
     }
-    return [
-        { tag: "h2", children: "Bobril-build 2" },
-        b.styledDiv(<b.IBobrilChildArray>[
-            s.disconnected ? "Disconnected" : s.connected ? "Connected" : "Connecting",
-            createCheckbox("Live reload", s.liveReload, com.setLiveReload),
-            s.actions.map((a) => [" ", createActionUI(a)])]),
-        getBuildStatus(),
-        getAgentsShort(selectedAgent, i => { selectedAgent = i; b.invalidate() }),
-        selectedAgent >= 0 && getAgentDetail(s.testSvrState.agents[selectedAgent])
-    ];
+    return b.createComponent({
```

Comment:
This call creates new component factory every time. It should be defined as a standalone component.